src/compiler/checker.ts

```diff
@@ -18359,9 +18359,9 @@ namespace ts {
                     if (relation === comparableRelation && target.flags & TypeFlags.Primitive) {
                         const constraints = sameMap((source as IntersectionType).types, t => t.flags & TypeFlags.Primitive ? t : getBaseConstraintOfType(t) || unknownType);
                         if (constraints !== (source as IntersectionType).types) {
-                            source = getIntersectionType(constraints);
-                            if (!(source.flags & TypeFlags.Intersection)) {
-                                return isRelatedTo(source, target, /*reportErrors*/ false);
+                            const newSource = getIntersectionType(constraints);

```

Comment:
These three lines should just be left unchanged. The intent is to overwrite the original intersection with the new and more specific intersection, and we rely on this substitution in the code the follows.